### ISSUE TITLE
Minibatches and KL re-weighting

### ISSUE THREAD

**Issue by ferrine** [pymc-devs/pymc#1287]
Hi! 
I wonder if you do KL re-weighting in elbo before taking gradients?
According to [this](http://arxiv.org/abs/1505.05424) paper it is better to optimize
(log p(w)  - log q(w|theta))/N_BATCHES + log p(BATCH|w) 
rather than 
log p(w)  - log q(w|theta) + log p(BATCH|w) 

BTW why do you support only adagrad for optimization?

**Comment by twiecki** [comment]
CC @taku-y

**Comment by taku-y** [comment]
Thanks for letting us know the paper. I think you said about Eq. (9), where minibatches are weighted based on the number of samples (minibatches) used for training. Although it might work, I need to think more about this. 

I think implementing optimization methods other than adagrad is more important. I will refactor ADVI functions to incorporate arbitrary SGD-based methods.

**Comment by ferrine** [comment]
That would be great!

**Comment by twiecki** [comment]
Partly closed by #1294. @ferrine If you want to give the weighting a try we would definitely consider if it worked well.

**Comment by ferrine** [comment]
Hi! 
The bad side of NOT re-weighting variational part of ELBO is that your training algorithm will be strongly affected by it. Prior will overwhelm data and you'll not get the true optima. Here is an illustration of this effect:

![scale_compare](https://cloud.githubusercontent.com/assets/11705326/18363926/3bf0c786-761d-11e6-8ed6-4ff9e09edd6a.png)
![scale_compare2](https://cloud.githubusercontent.com/assets/11705326/18364450/652cc904-761f-11e6-9fc9-c1e0411c38ca.png)

Here I used my old [wrapper](https://github.com/Lasagne/Lasagne/pull/731) for lasagne and your toy model and data specification from [here](http://twiecki.github.io/blog/2016/06/01/bayesian-deep-learning/). I was able to control every step of training. So I just specified `coef_sample = coef_scale = 0.05` as shared vars and then trained on sampled minibatches `U[0,1] < coef_sample`. I got the convergence and then said that `coef_scale = 1`, this moment can be easily detected on both plots. After some more steps I was sure that I've lost my optima, where accuracy was pretty well both on train and test, without any hope to get back.

This problem is very urgent since you cannot deal with large datasets that are trained with minibatches only. I hope the evidence is good enough to start refactoring code.

I have some kind of wishlist:
- Allow user to set `coef_scale` as shared var. As you can see above, it helped me a lot with experiments.
- Add possibility to call custom callbacks after every step of training loop. (ideally in simple advi too)
  One usage example. I want to control `coef_scale`: start with low so we see data a lot and go to the "good" model faster, but at every iteration we change `coef_scale` to be closer to `coef_sample` so that regularization will take affect when some preliminary steps are done. As a contrast you can see the paragraph after eq (9) [here](https://arxiv.org/pdf/1505.05424v2.pdf). They have similar trick but they regularize hard the beginning of the epoch and relax in the end.
- Separate loss creation step from advi function.
  It is a util that returns a tuple of `q(w)-p(w)` and `p(D|w)` so above wishes will be delegated to the user

**Comment by taku-y** [comment]
I don't understand the problem correctly, because there is no definition of `coef_scale` and `coef_sample`. Though, I think if `q(w)-p(w)` and `p(D|w)` can be obtained, you can customize the weightning algorithm. Is it right?

**Comment by ferrine** [comment]
@taku-y, 
`coef_sample` and `coef_scale` were my definitions for experiment, there are no equivalent in pymc3 so.

What about `q(w)-p(w)` and `p(D|w)` it will be tricky to implement some flexible interface since you have a some magic with result construction. I think that callback processing like

```
if callbacks:
    i = iteration_number
    l = int(uw_i.size / 2)
    u = bij.rmap(uw_i[:l])
    w = bij.rmap(uw_i[l:])
    for cb in callbacks:
        cb(i, l, u, w)
```

will be enough for user to do just anything
coef_scale can be implemented as additional kwarg for advi or advi_minibatch for example. If user passes a shared variable it will be possible for him to control weighting via callback

**Comment by ferrine** [comment]
Hi, what is the status of reweighting p(w) in advi minibatch?

**Comment by taku-y** [comment]
@ferrine I'm sorry for late to reply. I think it is not difficult to weightning log likelihood of prior distribution of model parameters, but difficult for variational distribution, because variational distribution is computed on concatenated random variable. I need to think about how log q(w) is computed in the code.

**Comment by taku-y** [comment]
I think we also need to write tests for reweighting, e.g., check the log probability of a model with reweighting. Do you have any ideas?

**Comment by ferrine** [comment]
We can use test value to check it. But I'm not sure it is easy to write.

**Comment by ferrine** [comment]
We can do it with analytical Ridge regression solution. Or we can just compare minibatch advi and simple advi

**Comment by taku-y** [comment]
I think ridge regression is good for test, thanks.

**Comment by springcoil** [comment]
@ twiecki 6@ 😁tutu5k 5k tutu5k I'm 5k 56 mims I'm in london south bank
holiday here tomorrow so 656 550k 655se 5555665

**Comment by ferrine** [comment]
Closed by #1684
